internal/pkg/test/tool/require/require_linux.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package require
+
+import (
+	"fmt"
+	"io"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	"github.com/sylabs/singularity/pkg/network"
+)
+
+var hasUserNamespace bool
+var hasUserNamespaceOnce sync.Once
+
+var supportNetwork bool
+var supportNetworkOnce sync.Once
+
+// UserNamespace checks that the current test could use
+// user namespace, if user namespaces are not enabled or
+// supported, the current test is skipped with a message.
+func UserNamespace(t *testing.T) {
+	// not performance critical, just save extra execution
+	// to get the same result
+	hasUserNamespaceOnce.Do(func() {
+		// user namespace is a bit special, as there is no simple
+		// way to detect if it's supported or enabled via a call
+		// on /proc/self/ns/user, the easiest and reliable way seems
+		// to directly execute a command by requesting user namespace
+		cmd := exec.Command("/bin/true")
+		cmd.SysProcAttr = &syscall.SysProcAttr{
+			Cloneflags: syscall.CLONE_NEWUSER,
+		}
+		// no error means user namespaces are enabled
+		hasUserNamespace = cmd.Run() == nil
+	})
+	if !hasUserNamespace {
+		t.Skipf("user namespaces seems not enabled or supported")
+	}
+}
+
+// Network check that bridge network is supported by
+// system, if not the current test is skipped with a
+// message.
+func Network(t *testing.T) {
+	supportNetworkOnce.Do(func() {
+		var err error
+		var cmdPath string
+		var stdinPipe io.WriteCloser
+
+		cmdPath, err = exec.LookPath("cat")
+		if err != nil {
+			t.Fatal(err)
+		}
+		cmd := exec.Command(cmdPath)
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+		cmd.SysProcAttr.Cloneflags = syscall.CLONE_NEWNET
+
+		stdinPipe, err = cmd.StdinPipe()
+		if err != nil {
+			return
+		}
+
+		if err := cmd.Start(); err != nil {
+			return
+		}
+
+		nsPath := fmt.Sprintf("/proc/%d/ns/net", cmd.Process.Pid)
```

Comment:
Why do we start the `cat` command here? Could we instead use `/proc/self/ns/net`? If not, why not?

Reply:
By doing that we would screw up the current network namespace which is the host namespace, running `cat` is a convenient way to pause (until we close stdin pipe) while we are manipulating the command process namespace from Go side without interfering with current namespaces.